hazelcast/src/hazelcast/client/impl/statistics/Statistics.cpp

```diff
@@ -0,0 +1,234 @@
+                                                                            logger(clientContext.getLogger()),
+                                                                            periodicStats(*this) {
+                    this->enabled = clientProperties.getStatisticsEnabled().getBoolean();
+                }
+
+                void Statistics::start() {
+                    if (!enabled) {
+                        return;
+                    }
+
+                    int64_t periodSeconds = clientProperties.getStatisticsPeriodSeconds().getLong();
+                    if (periodSeconds <= 0) {
+
+                        int64_t defaultValue = util::IOUtil::to_value<int64_t>(
+                                clientProperties.getStatisticsPeriodSeconds().getDefaultValue());
+                        logger.warning() << "Provided client statistics "
+                                         << clientProperties.getStatisticsPeriodSeconds().getName()
+                                         << " can not be less than or equal to 0. You provided " << periodSeconds
+                                         << " seconds as the configuration. Client will use the default value of "
+                                         << defaultValue << " instead.";
+                        periodSeconds = defaultValue;
+                    }
+
+                    schedulePeriodicStatisticsSendTask(periodSeconds);
+
+                    logger.info() << "Client statistics is enabled with period " << periodSeconds << " seconds.";
+
+                }
+
+                void Statistics::schedulePeriodicStatisticsSendTask(int64_t periodSeconds) {
+                    clientContext.getClientExecutionService().scheduleWithRepetition(
+                            boost::shared_ptr<util::Runnable>(new CollectStatisticsTask(*this)), 0,
+                            periodSeconds * 1000);
+                }
+
+                boost::shared_ptr<connection::Connection> Statistics::getOwnerConnection() {
+                    connection::ClientConnectionManagerImpl &connectionManager = clientContext.getConnectionManager();
+                    boost::shared_ptr<connection::Connection> connection = connectionManager.getOwnerConnection();
+                    if (NULL == connection.get()) {
+                        return boost::shared_ptr<connection::Connection>();
+                    }
+
+                    boost::shared_ptr<Address> ownerConnectionAddress = connectionManager.getOwnerConnectionAddress();
+                    int serverVersion = connection->getConnectedServerVersion();
+                    if (serverVersion < FEATURE_SUPPORTED_SINCE_VERSION) {
+                        // do not print too many logs if connected to an old version server
+                        const boost::shared_ptr<Address> ownerAddr = ownerAddress.get();
+                        if (ownerAddr.get() == NULL ||
```

Comment:
ownerConnectionAddress: The address got from Connection Manager.
ownerAddr: Address kept internally in the Statistics.

The code follows the same logic as in Java. If refactoring is needed, we need to do the same in Java as well, i am not sure if introducing the method is needed since it is just about log printing.

Variable names can be changed if you think it makes it easier to understand. I changed them.